app/models/post.rb

```diff
@@ -80,7 +80,28 @@ class Post < ActiveRecord::Base
   register_custom_field_type(MISSING_UPLOADS_IGNORED, :boolean)
 
   scope :private_posts_for_user, ->(user) {
-    where("posts.topic_id IN (#{Topic::PRIVATE_MESSAGES_SQL})", user_id: user.id)
+    joins("LEFT JOIN topic_allowed_users ON topic_allowed_users.topic_id = posts.topic_id AND topic_allowed_users.user_id = #{user.id.to_i}")
+      .joins("LEFT JOIN group_users ON group_users.user_id = #{user.id.to_i}")
+      .joins("LEFT JOIN topic_allowed_groups ON topic_allowed_groups.topic_id = posts.topic_id AND topic_allowed_groups.group_id = group_users.group_id")
+      .where("topic_allowed_users.topic_id IS NOT NULL OR topic_allowed_groups.topic_id IS NOT NULL")
+  }
+
+  scope :all_posts_for_user, ->(user) {
+    scope = joins(:post_search_data, :topic)
```

Comment:
Is `post_search_data` needed here? This seems like a search-specific thing so maybe it shouldn't be in the scope?

Reply:
Yea it is required because I’m joining now which does work with active record or.

Reply:
Where are we using activerecord `or`? I think this commit removes it from `search.rb`? 

My main worry is that we now have a scope called `Post.all_posts_for_user` which could be a little misleading because:
- It joins to `post_search_data`. Which makes sense for the search use case... but otherwise is a bit weird
- It checks access permissions for PMs, but does not check category/whisper permissions for regular topics. I worry someone might use this, thinking it's secure, and then we have a data leak

A more specific name might help
```
scope :insecure_posts_and_secure_private_messages_for_user_with_search_data
```
But clearly that's getting a bit ridiculous 😜 

Since this is search-specific, could we keep it in `search.rb` instead?